### PR TITLE
Introduce InstrumentationContext class

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do

# Motivation
https://github.com/DataDog/dd-trace-java/pull/5048#discussion_r1163283367

# Additional Notes

### CODE REVIEW

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java [hunk 1]
```diff
@@ -419,18 +420,20 @@ private InstrumentationResult applyInstrumentation(
     InstrumentationResult.Status status =
         preCheckInstrumentation(diagnostics, classLoader, methodNode);
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java:421]: Maybe InstrumentationContext can also be merged InstrumentationResult ? it only need Status.

Moreover, maybe we should put status inside InstrumentationContext and utility functions to addDiagnostic(DiagnosticMessage.Kind.ERROR, t), maybe that can also set the status to ERROR (but that might feel like dark magic)

which then we can also move preCheckInstrumentation into InsturmentationContext as well.

WDYT?

**Review by shatzi — APPROVED** [review]
it looks better IMHO

**Comment by richardstartin** [comment]
I just noticed this, there is already an `InstrumentationContext` class and overloading its name like this might create confusion for people talking about the architecture, people will have to say "Instrumentation context as in the debugger" or "Instrumentation context as in field injection" or actually say the package names when talking about the software.

**Comment by jpbempel** [comment]
will revisit this later

**Pull request closed without merging**